gitlab_test.go

```diff
@@ -0,0 +1,268 @@
+ }
+`
+
+	// https://gitlab.com/nakatanakatana/reviewdog/merge_requests/1
+	owner := "nakatanakatana"
+	repo := "reviewdog"
+	pr := 1
+	g, err := NewGitLabMergeReqest(client, owner, repo, pr, "")
+	if err != nil {
+		t.Fatal(err)
+	}
+	b, err := g.Diff(context.Background())
+	if err != nil {
+		t.Fatal(err)
+	}
+	if got := string(b); got != want {
+		t.Errorf("got:\n%v\nwant:\n%v", got, want)
+	}
+}
+
+func TestGitLabMergeRequest_comment(t *testing.T) {
+	if testing.Short() {
+		t.Skip("skipping test which contains actual API requests in short mode")
+	}
+	client := setupGitLabClient()
+	if client == nil {
+		t.Skip(notokenSkipTestMes)
+	}
+	// https://gitlab.com/nakatanakatana/reviewdog/merge_requests/1
+	owner := "nakatanakatana"
+	repo := "reviewdog"
+	pr := 1
+	g, err := NewGitLabMergeReqest(client, owner, repo, pr, "")
+	if err != nil {
+		t.Fatal(err)
+	}
+	comments, err := g.comment(context.Background())
+	if err != nil {
+		t.Fatal(err)
+	}
+	for _, c := range comments {
+		t.Log("---")
+		t.Log(c.Note)
+		t.Log(c.Path)
+		t.Log(c.Line)
+	}
+}
+
+func TestGitLabPullRequest_Post_Flush_review_api(t *testing.T) {
+	apiCalled := 0
+	mux := http.NewServeMux()
+	mux.HandleFunc("/projects/o/r/merge_requests/14/commits", func(w http.ResponseWriter, r *http.Request) {
+		apiCalled++
+		if r.Method != "GET" {
+			t.Errorf("unexpected access: %v %v", r.Method, r.URL)
+		}
+		cs := []*gitlab.Commit{
+			{
+				ID:      "0123456789abcdef",
+				ShortID: "012345678",
+			},
+		}
+		if err := json.NewEncoder(w).Encode(cs); err != nil {
+			t.Fatal(err)
+		}
+	})
+	mux.HandleFunc("/projects/o/r/repository/commits/0123456789abcdef/comments", func(w http.ResponseWriter, r *http.Request) {
+		apiCalled++
+		if r.Method != "GET" {
+			t.Errorf("unexpected access: %v %v", r.Method, r.URL)
+		}
+		cs := []*gitlab.CommitComment{
+			{
+				Path: "notExistFile.go",
+				Line: 1,
+				Note: bodyPrefix + "\nalready commented",
+			},
+		}
+		if err := json.NewEncoder(w).Encode(cs); err != nil {
+			t.Fatal(err)
+		}
+	})
+	mux.HandleFunc("/projects/o/r/repository/commits/sha/comments", func(w http.ResponseWriter, r *http.Request) {
+		apiCalled++
+		if r.Method != "POST" {
+			t.Errorf("unexpected access: %v %v", r.Method, r.URL)
+		}
+		var req gitlab.CommitComment
+		if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
+			t.Error(err)
+		}
+		want := gitlab.CommitComment{
+			Path:     "notExistFile.go",
+			Line:     14,
+			Note:     bodyPrefix + "\nnew comment",
+			LineType: "new",
+		}
+		if diff := pretty.Compare(want, req); diff != "" {
+			t.Errorf("req.Comments diff: (-got +want)\n%s", diff)
+		}
+		if err := json.NewEncoder(w).Encode(req); err != nil {
+			t.Fatal(err)
+		}
+	})
+	ts := httptest.NewServer(mux)
+	defer ts.Close()
+
+	cli := gitlab.NewClient(nil, "")
+	cli.SetBaseURL(ts.URL)
+	g, err := NewGitLabMergeReqest(cli, "o", "r", 14, "sha")
+	if err != nil {
+		t.Fatal(err)
+	}
+	// Path is set to notExistFile path for mock-up test.
```

Comment:
How about "Path is set to non existing file path for mock test not to use last commit id of the line."?